### PR TITLE
[Doc] Add performance.rst introducing the 'ti.block_dim' API

### PR DESCRIPTION
**Pull request by archibate**
Related issue = https://forum.taichi.graphics/t/how-to-prevent-3d-mpm-from-becoming-memory-bound/1086/3

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
Amazing! `ti.block_dim(n_grid)` yields ~3x performance boost in `mpm3d.py` on my end! So I'd like share this with our users :)

### CODE REVIEW

File: docs/performance.rst [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+   For performance profiling utilities, see :ref:`profiler`.
+
+Hierarchical structure of GPU
+*****************************
+
+Taichi programs could be executed in parallel on GPU.
+The level of structure of GPU is defined hierarchically.
+
+From small to large, the computation units are:
+**invocation** < **thread** < **block** < **grid**.
```
Comment (yuanming-hu) [inline docs/performance.rst:29]: I think **iteration** is a better name compared to **invocation**. Please consider changing the word.

File: docs/performance.rst [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  All threads within a block are executed in **parallel**.
+  Threads within the same block can share their **block local storage**.
+
+- **grid**:
+  Blocks are grouped into grids.
+  Grid is the minimal unit that being **launched** from host.
+  All blocks within a grid are executed in **parallel**.
+  In Taichi, each **parallelized for-loop** is a grid.
+
+For more details, please see `this tutorial by CUDA official <http://users.wfu.edu/choss/CUDA/docs/Lecture%205.pdf>`_.
```
Comment (yuanming-hu) [inline docs/performance.rst:52]:
```suggestion
For more details, please see `the CUDA C programming guide <https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#thread-hierarchy>`_. The OpenGL and Metal backends follow a similar thread hierarchy.
```

File: docs/performance.rst [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+.. note::
+
+   For performance profiling utilities, see :ref:`profiler`.
+
+Hierarchical structure of GPU
+*****************************
+
+Taichi programs could be executed in parallel on GPU.
+The level of structure of GPU is defined hierarchically.
```
Comment (yuanming-hu) [inline docs/performance.rst:26]:
```suggestion
GPUs have a **thread hierarchy**.
```

File: docs/performance.rst [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+This allows advanced users to manually fine tune the performance to be better.
+
+For example, specifying a suitable ``ti.block_dim`` could yield almost 3x performance boost in `examples/mpm3d.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm3d.py>`_!
+
+.. note::
+
+   For performance profiling utilities, see :ref:`profiler`.
+
+Hierarchical structure of GPU
+*****************************
```
Comment (yuanming-hu) [inline docs/performance.rst:23]:
```suggestion
Thread hierarchy of GPUs
************************
```

File: docs/performance.rst (hunk 1 above)
Comment (archibate) [reply docs/performance.rst:29]:
```suggestion
**iteration** < **thread** < **block** < **grid**.
```

File: docs/performance.rst [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+From small to large, the computation units are:
+**invocation** < **thread** < **block** < **grid**.
+
+- **invocation**:
+  Invocation is the **body of a for-loop**.
+  Each invocation corresponding to a specific ``i`` value in for-loop.
+
+- **thread**:
+  Invocations are grouped into threads.
```
Comment (archibate) [inline docs/performance.rst:36]:
```suggestion
- **iteration**:
  Iteration is the **body of a for-loop**.
  Each iteration corresponding to a specific ``i`` value in for-loop.

- **thread**:
  Iterations are grouped into threads.
```

File: docs/performance.rst [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  Threads are the minimal unit that is parallelized.
+  All invocations within a thread are executed in **serial**.
+  We usually use 1 invocation per thread for maximizing parallel performance.
```
Comment (archibate) [inline docs/performance.rst:39]:
```suggestion
  All iterations within a thread are executed in **serial**.
  We usually use 1 iteration per thread for maximizing parallel performance.
```

File: docs/performance.rst [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+    Having a large ``thread_dim`` could be helpful when your program have
+    reduction over a single variable.
+    Usually the Taichi compiler will choose a optimal ``thread_dim`` when
+    such reduction is detected.
+    But you may fine tune it manually to get the best performance.
+
+    .. note::
+
+        This function is only available on OpenGL for now.
```
Comment (archibate) [inline docs/performance.rst:108]:
```suggestion
```

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 8]
```diff
@@ -708,7 +708,7 @@ class KernelGen : public IRVisitor {
   }
 
   void generate_grid_stride_loop_header() {
-    ScopedIndent _s(line_appender_);
+    ScopedIdent _s(line_appender_);
```
[the remark this answers is not in the document]
Comment (archibate) [reply taichi/backends/opengl/codegen_opengl.cpp:711]:
```suggestion
    ScopedIndent _s(line_appender_);
```

File: docs/performance.rst [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+For-loop decorators
+-------------------
+
+In Taichi, as we all know, the outermost for-loop is parallelized.
+
+However, there're some implementation details about **how it is parallelized**.
+
+Taichi provides some API to modify these parameters which is usually invisible
+from end-users.
+This allows advanced users to manually fine tune the performance to be better.
```
Comment (yuanming-hu) [inline docs/performance.rst:14]:
```suggestion
This allows advanced users to manually fine-tune the performance.
```

File: docs/performance.rst [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+For example, specifying a suitable ``ti.block_dim`` could yield almost 3x performance boost in `examples/mpm3d.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm3d.py>`_!
```
Comment (yuanming-hu) [inline docs/performance.rst:16]:
```suggestion
For example, specifying a suitable ``ti.block_dim`` could yield an almost 3x performance boost in `examples/mpm3d.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm3d.py>`_.
```

File: docs/performance.rst [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+  Grid is the minimal unit that being **launched** from host.
+  All blocks within a grid are executed in **parallel**.
+  In Taichi, each **parallelized for-loop** is a grid.
+
+For more details, please see `this tutorial by CUDA official <http://users.wfu.edu/choss/CUDA/docs/Lecture%205.pdf>`_.
+
+API reference
+*************
+
+We may **prepend** some decorator(s) to tweak the property of a for-loop, e.g.:
```
Comment (yuanming-hu) [inline docs/performance.rst:56]:
```suggestion
Programmers may **prepend** some decorator(s) to tweak the property of a for-loop, e.g.:
```

File: docs/performance.rst [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+
+Here's the list of available decorators:
+
+
+.. function:: ti.block_dim(n)
+
+    :parameter n: (int) threads per block / block dimension
+
+    Specify the **threads per block** of next parallelized for-loop.
```
Comment (yuanming-hu) [inline docs/performance.rst:80]:
```suggestion
    Specify the **threads per block** of the next parallel for-loop.
```

**Review by yuanming-hu — APPROVED** [review]
Thanks! I took another pass. This mostly looks good to me now.

**Pull request merged**